spec/unit/face/catalog_spec.rb

```diff
@@ -0,0 +1,39 @@
+require 'spec_helper'
+require 'puppet/face'
+require 'puppet/indirector/facts/facter'
+require 'puppet/indirector/facts/rest'
+
+describe Puppet::Face[:catalog, '0.0.1'] do
+
+  describe '#download' do
+    let(:model) { Puppet::Node::Facts }
+    let(:test_data) { model.new('puppet.node.test', {test_fact: 'catalog_face_request_test_value'}) }
+    let(:catalog) { Puppet::Resource::Catalog.new('puppet.node.test', Puppet::Node::Environment.remote(Puppet[:environment].to_sym)) }
+
+    before(:each) do
+      Puppet[:facts_terminus] = :memory
+      Puppet::Node::Facts.indirection.save(test_data)
+      allow(Puppet::Face[:catalog, "0.0.1"]).to receive(:save).once
+
+      Puppet.settings.parse_config(<<-CONF)
+[main]
+server=puppet.server.test
+certname=puppet.node.test
+CONF
+
+      # Faces start in :user run mode
+      Puppet.settings.preferred_run_mode = :user
+    end
+
+    it "adds facts to the catalog request" do
+      stub_request(:post, 'https://puppet.server.test:8140/puppet/v3/catalog/puppet.node.test?environment=*root*')
+        .with(
+          headers: { 'Content-Type' => 'application/x-www-form-urlencoded' },
+          body: /catalog_face_request_test_value/
```

Comment:
I ended up using a regex here because I couldn't figure out how to use `hash_including` when the `facts` value is a url double-encoded.

Reply:
IIRC the double encoded facts are added as the value of the `facts` query parameter, so you can do:

https://github.com/puppetlabs/puppet/blob/d6151b2135d5bb68759d3bc878b07ea5b59ee00c/spec/unit/http/service/compiler_spec.rb#L262